peps/pep-0771.rst

```diff
@@ -0,0 +1,502 @@
+    ]
+
+and:
+
+.. code-block:: toml
+
+    [project]
+    default-optional-dependencies = [
+        "backend1",
+        "backend2",
+        "backend3"
+    ]
+
+
+Overriding default extras
+-------------------------
+
+If extras are explicitly given in a dependency specification, the default
+extras are not installed. Otherwise, the default extras are used.
+
+For example, if a package
+defines a ``extra1`` default extra as well as a non-default ``extra2``
+extra, then if a user was to install the package with::
+
+    pip install package
+
+the ``extra1`` dependency would be included. If the user instead uses::
+
+    pip install package[extra2]
+
+then the ``extra1`` extras would not be installed.
+
+If the same package is specified multiple times in a dependency tree, then if
+any of these were specified without any explicit extras, the default extras
+should be installed even if other occurrences of the package in the dependency
+tree have explicitly specified extras.
+
+Note that ``package[]`` would continue to be equivalent to ``package`` and would
+not be provided as a way to install without default extras (see the `Rejected
+Ideas`_ section for the rationale).
+
+Installing without default extras
+---------------------------------
+
+In some cases, package maintainers may want to facilitate installing packages
+without any default extras. In this case, as will be shown in more detail in
+`How to teach this`_, the best approach is to define an extras which could be
+called e.g. ``minimal`` or ``nodefault`` (the naming would be up to the package
+maintainer) which would be an empty set of dependencies. If this extras is
+specified, no default extras will be included, so that e.g. ``package[minimal]``
+would include only required dependencies and no extras. Note that this requires
+no additional specification and is a natural consequence of the rule described
+in `Overriding default extras`_.
+
+There are however valid use cases where package maintainers may not want to
+provide this. For example, in the case of the multiple possible frontends or
+backends, it may be that the package would not be functional without any of the
+options. To take a specific example, a package may need either PyQt or PySide to
+be installed but will not work if none are provided, so a package maintainer may
+therefore not want to provide an option to install the package without any
+extras.
+
+Backward Compatibility
+======================
+
+All package specification cases valid under :pep:`508` will remain valid.
+Therefore, this proposal is fully backward-compatible with existing :pep:`508`
+usage.
+
+Once packages start defining default extras, those defaults will only be honored
+with recent versions of packaging tools which implement this PEP, but those
+packages will remain fully backward-compatible with older packaging tools - with
+the only difference that the default extras will not be installed automatically
+for older packaging tools.
```

Comment:
How about: "... will not be installed automatically when older packaging tools are used." ?

Reply:
Done!